win32/src/job_object_helper.c

```diff
@@ -11,7 +11,130 @@
+                        result = MU_FAILURE;
+                    }
+                    else
+                    {
+                        /*Codes_SRS_JOB_OBJECT_HELPER_18_010: [ job_object_helper_limit_resources shall call SetInformationJobObject, passing JobObjectExtendedLimitInformation and a JOBOBJECT_EXTENDED_LIMIT_INFORMATION object with JOB_OBJECT_LIMIT_JOB_MEMORY set and JobMemoryLimit set to the percent_physical_memory percent of the physical memory in bytes. ]*/
+                        JOBOBJECT_EXTENDED_LIMIT_INFORMATION extended_limit_information = {0};
+                        extended_limit_information.BasicLimitInformation.LimitFlags  = JOB_OBJECT_LIMIT_JOB_MEMORY;
+                        extended_limit_information.JobMemoryLimit = percent_physical_memory * memory_status_ex.ullTotalPhys / 100;
+                        if (!SetInformationJobObject(job_object, JobObjectExtendedLimitInformation, &extended_limit_information, sizeof(extended_limit_information)))
+                        {
+                            LogLastError("SetInformationJobObject("
+                                " job_object = %p"
+                                ", JobObjectExtendedLimitInformation"
+                                ", &extended_limit_information = %p"
+                                ", sizeof(extended_limit_information) = %zu) failed",
+                                job_object,
+                                &extended_limit_information,
+                                sizeof(extended_limit_information));
+                            /*Codes_SRS_JOB_OBJECT_HELPER_18_015: [ If there are any failures, job_object_helper_limit_resources shall fail and return a non-zero value. ]*/
+                            result = MU_FAILURE;
+                        }
+                        else
+                        {
+                            /*Codes_SRS_JOB_OBJECT_HELPER_18_011: [ job_object_helper_limit_resources shall call SetInformationJobObject passing JobObjectCpuRateControlInformation and a JOBOBJECT_CPU_RATE_CONTROL_INFORMATION object with JOB_OBJECT_CPU_RATE_CONTROL_ENABLE and JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP set, and CpuRate set to percent_cpu times 100. ]*/
+                            JOBOBJECT_CPU_RATE_CONTROL_INFORMATION cpu_rate_control_information = {0};
+                            cpu_rate_control_information.ControlFlags = JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP;
+                            cpu_rate_control_information.CpuRate = percent_cpu * 100;
+                            if (!SetInformationJobObject(job_object, JobObjectCpuRateControlInformation, &cpu_rate_control_information, sizeof(cpu_rate_control_information)))
+                            {
+                                LogLastError("SetInformationJobObject("
```

Comment:
>LogLastError [](http://example.com/codeflow?start=32&length=12)

assume line 105 gets executed. That implies that line 84 succeeded and did set the memory limit. However, the CPU limit failed.

At this moment I believe that either: the RAM should be reverted to "what it used to be before" or some other remedies.

Reply:
Good idea, and this wouldn't be an issue if it were 2 separate APIs. I like it, but I don't want to change it now because of time.